flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java

```diff
@@ -0,0 +1,197 @@
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.scheduler.DeploymentOption;
+import org.apache.flink.runtime.scheduler.ExecutionVertexDeploymentOption;
+import org.apache.flink.runtime.scheduler.SchedulerOperations;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link SchedulingStrategy} instance for batch job which schedule vertices when input data are ready.
+ */
+public class LazyFromSourcesSchedulingStrategy implements SchedulingStrategy {
+
+	private final SchedulerOperations schedulerOperations;
+
+	private final SchedulingTopology schedulingTopology;
+
+	private final Map<ExecutionVertexID, DeploymentOption> deploymentOptions;
+
+	private final InputDependencyConstraintChecker inputConstraintChecker;
+
+	public LazyFromSourcesSchedulingStrategy(
+			SchedulerOperations schedulerOperations,
+			SchedulingTopology schedulingTopology) {
+		this.schedulerOperations = checkNotNull(schedulerOperations);
+		this.schedulingTopology = checkNotNull(schedulingTopology);
+		this.deploymentOptions = new HashMap<>();
+		this.inputConstraintChecker = new InputDependencyConstraintChecker();
+	}
+
+	@Override
+	public void startScheduling() {
+		List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new ArrayList<>();
+		final DeploymentOption updateOption = new DeploymentOption(true);
+		final DeploymentOption nonUpdateOption = new DeploymentOption(false);
+
+		for (SchedulingExecutionVertex schedulingVertex : schedulingTopology.getVertices()) {
+			DeploymentOption option = nonUpdateOption;
+			for (SchedulingResultPartition srp : schedulingVertex.getProducedResultPartitions()) {
+				if (srp.getPartitionType().isBlocking()) {
+					inputConstraintChecker.addSchedulingResultPartition(srp);
+				} else {
+					option = updateOption;
+				}
+			}
+			deploymentOptions.put(schedulingVertex.getId(), option);
+
+			if (schedulingVertex.getConsumedResultPartitions().isEmpty()) {
+				// schedule vertices without consumed result partition
+				executionVertexDeploymentOptions.add(
+						new ExecutionVertexDeploymentOption(schedulingVertex.getId(), option));
+			}
+		}
+
+		schedulerOperations.allocateSlotsAndDeploy(executionVertexDeploymentOptions);
+	}
+
+	@Override
+	public void restartTasks(Set<ExecutionVertexID> verticesToRestart) {
+		// increase counter of the dataset first
+		for (ExecutionVertexID executionVertexId : verticesToRestart) {
+			final SchedulingExecutionVertex schedulingVertex = schedulingTopology.getVertex(executionVertexId)
+					.orElseThrow(() -> new IllegalStateException("can not find scheduling vertex for " + executionVertexId));
+
+			for (SchedulingResultPartition srp : schedulingVertex.getProducedResultPartitions()) {
+				if (srp.getPartitionType().isBlocking()) {
```

Comment:
Do you think the code can be simplified if we track the intermediate data sets for pipelined partitions? I would be willing to pay the extra memory costs.

Reply:
OK